spikeinterface/comparison/addtemplatesrecording.py

```diff
@@ -0,0 +1,130 @@
+from typing import List, Union
+import numpy as np
+from spikeinterface.core import BaseRecording, BaseRecordingSegment, BaseSorting, BaseSortingSegment
+
+
+class AddTemplatesRecording(BaseRecording):
+    """
+    Class for creating a recording based on spike timings and templates.
+    Can be just the templates or can add to an already existing recording.
+
+    Parameters
+    ----------
+    sorting: BaseSorting
+        Sorting object containing all the units and their spike train.
+    templates: np.ndarray[n_units, n_samples, n_channels]
+        Array containing the templates to inject for all the units.
+    nbefore: list[int] | int | None
+        Where is the center of the template for each unit?
+        If None, will default to the highest peak.
+    amplitude_factor: list[list[float]] | int | None
+        The amplitude of each spike for each unit (1.0=default).
+        If None, will default to 1.0 everywhere.
+    target_recording: BaseRecording | None
+        The recording over which to add the templates.
+        If None, will default to traces containing all 0.
+    t_max: list[int] | None
+        The number of frames in the recording.
+    """
+
+    def __init__(self, sorting: BaseSorting, templates: np.ndarray, nbefore: Union[List[int], int, None] = None,
+                 amplitude_factor: Union[List[List[float]], None] = None,
+                 target_recording: Union[BaseRecording, None] = None, t_max: Union[List[int], None] = None) -> None:
+        channel_ids = target_recording.channel_ids if target_recording is not None else list(range(templates.shape[2]))
+        BaseRecording.__init__(self, sorting.get_sampling_frequency(), channel_ids, templates.dtype)
+        
+        n_units = len(sorting.unit_ids)
+        assert len(templates) == n_units
+        self.spike_vector = sorting.to_spike_vector()
+
+        if nbefore is None:
+            nbefore = np.argmax(np.max(np.abs(templates), axis=2), axis=1)
+        elif isnumeric(nbefore):
+            nbefore = [nbefore]*n_units 
+        else:
+            assert len(nbefore) == n_units
+
+        if amplitude_factor is None:
+            amplitude_factor = [[1.0]*len(sorting.get_unit_spike_train(unit_id)) for unit_id in sorting.unit_ids]
+
+        if target_recording is not None:
+            assert target_recording.get_num_segments() == sorting.get_num_segments()
+            target_recording.copy_metadata(self)
+
+        if t_max is None:
+            if target_recording is None:
+                t_max = [self.spike_vector['sample_ind'][-1] + templates.shape[1]]
+            else:
+                t_max = [target_recording.get_num_frames(segment_index) for segment_index in range(sorting.get_num_segments())]
+
+
+        for segment_index in range(sorting.get_num_segments()):
+            spikes = self.spike_vector[self.spike_vector['segment_ind'] == segment_index]
```

Comment:
Maybe a numpy.searchsorted would be faster here.
Keep in mind that this unit is call several when we use n_jobs>1